src/odbc/src/odbc.cpp

```diff
@@ -1182,7 +1185,18 @@ SQLRETURN SQLGetData(SQLHSTMT stmt, SQLUSMALLINT colNum, SQLSMALLINT targetType,
   ApplicationDataBuffer dataBuffer(driverType, targetValue, bufferLength,
                                    strLengthOrIndicator);
 
+  // Cell offsets are used between SQLGetData calls to return
+  // variable-length data in parts. Statement keeps the cell offset
+  // between repeated calls on the same column.
+  if (colNum != statement->GetCurrentColNum()) {
+    statement->SetCurrentColNum(colNum);
+    statement->SetCellOffset(0);
+    dataBuffer.SetCellOffset(0);
+  } else {
+    dataBuffer.SetCellOffset(statement->GetCellOffset());
+  }
   statement->GetColumnData(colNum, dataBuffer);
+  statement->SetCellOffset(dataBuffer.GetCellOffset());
```

Comment:
This is how `SQLGetData` can return variable-length data in parts. `Statement` persists across `SQLGetData` calls, `dataBuffer` does not. The cell offset is used within `dataBuffer` to calculate the number of bytes to skip in a field when writing to a buffer.

Please let me know if you have a better way of handling this. In its current form, The application data buffer's cell offset must be reset for every row fetch. This means tests using application data buffers have to call buffer.SetCellOffset(0) after every call involving putting a string to a buffer. This is not ideal.

Reply:
Maybe a better way to do it would be with a pointer to a `Statement` cell offset. A new application data buffer constructor could be added like:
```
  ApplicationDataBuffer dataBuffer(driverType, targetValue, bufferLength,
                                   strLengthOrIndicator, &cellOffset);
```
Outside of its use in `SQLGetData`, the `cellOffsetPtr` in `ApplicationDataBuffer` could remain `nullptr`.

Reply:
I have gone with a method where `cellOffset` is only used by `ApplicationDataBuffer` during `SQLGetData`. It is initialized to `-1` to indicate that it shouldn't be used.

`Statement` now only has to manage the cell offset in `SQLGetData` and in `InternalFetchRow`, resetting `cellOffset` to `0` for each new row.